prow/test/integration/fakeghserver/fakeghserver.go

```diff
@@ -0,0 +1,219 @@
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/go-github/github"
+	"github.com/sirupsen/logrus"
+	prowgh "k8s.io/test-infra/prow/github"
+
+	"k8s.io/test-infra/prow/github/fakegithub"
+	"k8s.io/test-infra/prow/interrupts"
+	"k8s.io/test-infra/prow/logrusutil"
+	"k8s.io/test-infra/prow/pjutil"
+)
+
+var ()
+
+type options struct {
+	port int
+}
+
+func (o *options) validate() error {
+	return nil
+}
+
+func flagOptions() *options {
+	o := &options{}
+	flag.IntVar(&o.port, "port", 8888, "Port to listen on.")
+	return o
+}
+
+func main() {
+	logrusutil.ComponentInit()
+
+	o := flagOptions()
+	flag.Parse()
+	if err := o.validate(); err != nil {
+		logrus.WithError(err).Fatal("Invalid arguments.")
+	}
+	defer interrupts.WaitForGracefulShutdown()
+	ghClient := fakegithub.NewFakeClient()
+
+	mux := http.NewServeMux()
+	// setup common handlers for local and deployed runs
+	mux.Handle("/", reposHandler(ghClient))
+	mux.Handle("/fakeghserver/", http.StripPrefix("/fakeghserver", reposHandler(ghClient)))
+
+	health := pjutil.NewHealth()
+	health.ServeReady()
+
+	logrus.Info("Start server")
+
+	// setup done, actually start the server
+	server := &http.Server{Addr: ":8888", Handler: mux}
+	interrupts.ListenAndServe(server, 5*time.Second)
+}
+
+func reposHandler(ghc *fakegithub.FakeClient) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		redirect(w, r, ghc)
+	})
+}
+
+func unmarshal(w http.ResponseWriter, r *http.Request, data interface{}) error {
+	d := json.NewDecoder(r.Body)
+	d.DisallowUnknownFields()
+
+	w.Header().Set("Content-Type", "application/json")
+	err := d.Decode(&data)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "{\"error\": \"Failed unmarshal request: %v\"}", err.Error())
+	}
+	return err
+}
+
+// So far, supports APIs used by crier:
+//type GitHubClient interface {
+// 	BotName() (string, error) # /user
+// 	CreateStatus(org, repo, ref string, s github.Status) error # fmt.Sprintf("/repos/%s/%s/statuses/%s", org, repo, SHA)
+// 	ListIssueComments(org, repo string, number int) ([]github.IssueComment, error) # fmt.Sprintf("/repos/%s/%s/issues/%d/comments", org, repo, number)
+// 	CreateComment(org, repo string, number int, comment string) error # fmt.Sprintf("/repos/%s/%s/issues/%d/comments", org, repo, number),
+// 	DeleteComment(org, repo string, ID int) error # fmt.Sprintf("/repos/%s/%s/issues/comments/%d", org, repo, number),
+// 	EditComment(org, repo string, ID int, comment string) error # fmt.Sprintf("/repos/%s/%s/issues/comments/%d", org, repo, number),
+func redirect(w http.ResponseWriter, r *http.Request, ghc *fakegithub.FakeClient) error {
+	logrus.Infof("Got request %+v", *r)
+	var ok int
+	// By default ok is 200
+	ok = http.StatusOK
+	var err error
+	var msg string
+	parts := strings.Split(r.URL.Path, "/")
+	if strings.TrimSpace(parts[0]) == "" {
+		parts = parts[1:]
+	}
+	switch {
```

Comment:
can we just use a http.ServeMux for the routing rather than implementing it with 200 lines of nested switch statements that will only grow bigger over time?

Reply:
(if http.ServeMux lacks features, please look for another pre-defined router. Go is famous for having a not so powerfull router in the stdlib and plenthoa of OSS ones to make up for that)

Reply:
Updated to use routing from gorilla.mux